src/sage/libs/gap/libgap.pyx

```diff
@@ -160,112 +160,117 @@ using the recursive expansion of the
-   reinstall the original handler when the GAP code has finished. This
-   is accomplished using the suggestively-named ``gap_sig_on()`` and
-   ``gap_sig_off()`` functions. After you have called
-   ``gap_sig_on()``, if GAP receives Ctrl-C, it will invoke our custom
-   ``error_handler()`` that will set a :exc:`KeyboardInterrupt`
-   containing the phrase "user interrupt". Eventually (as explained in
-   the preceding paragraphs), control will jump back to the Cython
-   wrapper for ``GAP_Enter()``, and this exception will be raised.
-
-   Before you attempt to change any of this, please make sure that you
-   understand the issues that it is intended to fix, e.g.
-
-     * https://github.com/sagemath/sage/issues/37026
-     * https://trofi.github.io/posts/312-the-sagemath-saga.html
-     * https://github.com/sagemath/sage/pull/40585
-     * https://github.com/sagemath/sage/pull/40594
-     * https://github.com/sagemath/sage/issues/40598
+    We are using the GAP API provided by the GAP project since
+    GAP 4.10.
+
+    Calls to the GAP C library (functions declared in libgap-api.h)
+    should be sandwiched between calls to ``GAP_Enter()`` and
+    ``GAP_Leave()``. These are macros defined in ``libgap-api.h`` and
+    must be used carefully because ``GAP_Enter()`` is defined as two
+    function calls in succession without braces. The first thing that
+    ``GAP_Enter()`` does is a ``setjmp()`` which plays an important
+    role in handling errors. The return value from ``GAP_Enter()`` is
+    non-zero (success) the first time around, and if an error occurs,
+    execution "jumps" back to ``GAP_Enter()``, this time with a return
+    value of zero (failure). Due to these quirks, naive attempts to
+    handle the return value of ``GAP_Enter()`` are doomed to fail.
+    The correct pattern to use is,
+
+        try:
+            GAP_Enter()
+            # further calls to libgap
+        finally:
+            GAP_Leave()
+
+    How this works is subtle. When GAP is initialized, we install an
+    ``error_handler()`` callback that GAP invokes on error. This
+    function sets a python exception using ``PyErr_Restore()``, but so
+    long as we remain in C, this exception will not actually be
+    raised. When ``error_handler()`` finishes executing, control
+    returns to GAP which then jumps back to the previous
+    ``GAP_Enter()``. It is at this point that we need to raise the
+    (already set) exception, to prevent re-executing the code that
+    caused an error. To facilitate this, ``GAP_Enter()`` is wrapped by
+    Cython, and the wrapper is qualified with ``except 0``. This tells
+    Cython to treat a return value of zero as an error, and raise an
+    exception if an exception is set. (One will be set if there was an
+    error because our ``error_handler()`` sets it). Here is a real
+    example::
+
+        cpdef void crash_and_burn() except *:
+            cdef GapElement x = <GapElement>libgap({'a': 1, 'b': 2})
```

Comment:
is this dangerous? The Python wrapper `libgap(…)` is destroyed while the `GapElement` lives on.